lib/stdlib/src/queue.erl

```diff
@@ -453,6 +454,93 @@ fold(Fun, Acc0, {R, F}) when is_function(Fun, 2), is_list(R), is_list(F) ->
 fold(Fun, Acc0, Q) ->
     erlang:error(badarg, [Fun, Acc0, Q]).
 
+%% Check if any item satisfies the predicate, traverse in queue order.
+%%
+%% O(len(Q)) worst case
+-spec any(Pred, Q :: queue(Item)) -> boolean() when
+      Pred :: fun((Item) -> boolean()).
+any(Pred, {R, F}) when is_function(Pred, 1), is_list(R), is_list(F) ->
+    lists:any(Pred, F) orelse
+    lists:any(Pred, lists:reverse(R));
+any(Pred, Q) ->
+    erlang:error(badarg, [Pred, Q]).
+
+%% Check if all items satisfy the predicate, traverse in queue order.
+%%
+%% O(len(Q)) worst case
+-spec all(Pred, Q :: queue(Item)) -> boolean() when
+      Pred :: fun((Item) -> boolean()).
+all(Pred, {R, F}) when is_function(Pred, 1), is_list(R), is_list(F) ->
+    lists:all(Pred, F) andalso
+    lists:all(Pred, lists:reverse(R));
```

Comment:
`all_r(Pred, R);`

``` Erlang
all_r(Pred, [X|R]) when is_list(R) ->
    all_r(Pred, R) andalso Pred(X);
all_r(_Pred, []) ->
    true.
```
Not tested, nor compiled.  Does not build any garbage like `lists:reverse(R)`.

Reply:
Why the double negation `(not not Pred(X))`?

Reply:
Please reconsider, I didn't use `lists:reverse/1` here out of sheer lazyness ;)

(But first, I have to admit that I don't know how `lists:reverse/1` is really implemented. In any case, it is _fast_.)

As I see it, the difference between the two approaches is this:
* My approach only traverses the (reversed) `R` list down to the point where `Pred` returns `false`, then stops. In the worst case, when `Pred` returns `true` for all (or all but the last) elements in `R`, that means a full traversal of `R`, all other cases stop earlier. So, this approach is `O(length(R))` in the worst case (plus a bit for the reversal, yes ;)).
* The suggested approach traverses `R` in full _always_, then unrolls the recursion again. With this, even the best case (`Pred` returns `false` for the last element in `R`) would equal the worst case in my approach, plus the unrolling of the _entire_ recursion. So, this approach is `O(2*length(R))` in _any_ case.

The same holds for `any` (and actually, for `delete_rear` too, which could be implemented by ways of `lists:reverse` + `lists:delete` + `lists:reverse`, but since it involves two reversals, I refrained from it there).

Reply:
> Why the double negation `(not not Pred(X))`?

Because for the last iteration, `Pred`, that is user supplied, might return something non-boolean, which would become the return value of the function.  `(not not)` fails for a non-booelan operand, while `andalso` | `orelse` ignores their right operand.  But it is redundant to have it on every iteration since the result is recursed up to the first operand of `andalso` | `orelse`, so the better code would be:
``` Erlang
    not not all_r(Pred, R);

all_r(Pred, [X|R]) ->
    all_r(Pred, R) andalso Pred(X);
all_r(_Pred, []) ->
    true.
```

Reply:
> Please reconsider, I didn't use lists:reverse/1 here out of sheer lazyness ;)

The suggested approach doesn't create any garbage which is a significant win for processes with large heaps.

Why not leave the order undefined and just traverse it as-is? Honoring side-effects in _predicate functions_ feels rather strange and neither "all" nor "any" imply order per se.

Reply:
> Please reconsider, I didn't use `lists:reverse/1` here out of sheer lazyness ;)

I am sorry that I suspected that :-)

`lists:reverse/1` is implemented in C and it is as you say _fast_, but it is still O(`length(R)`).  It builds a complete reversed list on the heap, so it traverses all elements of `R`.  Then your iteration goes tail recursively down to a bad `Pred(X)` value and stops at that.  So it builds the whole reverse `R` on the heap, which is one list cell per element.  This has to be garbage collected later.

My suggestion iterates body recursively down to the end of the list, which builds stack, one return address per iteration.  Then it applies `Pred(X)` to the list elements up to a bad value and skips all after.  This just builds stack during the iteration, and it should be one word (half a list cell) per iteration, so it should be the half amount of memory.  The stack is free immediately after the iteration is complete without garbage collection.

Now, garbage collection happens anyway, so the reversing approach probably just makes it happen earlier than the body recursion approach.  Then the garbage collection time is mostly depending on the amount of live data.  So if this function is called from a process with much live data it might get an expensive garbage collection earlier.  The body recursion approach on the other hand has to do one `andalso` | `orelse` test per iteration even after `Pred(X)` has returned a bad value, while the reversing approach can finish right there.  `lists:reverse/1` is probably much faster than an erlang loop testing a boolean for the same list length.

So which is fastest depends on stuff.  Both are O(`length(R)`) (since garbage collection is not O(ProducedGarbage)).  Note that both O(2*N) and O(0.5*N) are O(N).  We are just playing with constant factors here.  The reversing approach is probably faster (since `lists:reverse/1` has so low constant factor on O(N)), at least when called from a process with small heap (say benchmark, and the question is how large heap that is needed to be noticed).  The body recursion approach probably uses less maximum memory.

Either approach is good, and I guess you have benchmarked.  I just like body recursion, and enjoy minimizing garbage. ;-)

Reply:
> @jhogberg asked a very good question kind of hidden in the buzz above so I repeat it here:
> 
> > Why not leave the order undefined and just traverse it as-is? Honoring side-effects in _predicate functions_ feels rather strange and neither "all" nor "any" imply order per se.
> 
> `lists:any/2` and `lists:all/2` does not guarantee anything about evaluation order. It is just assumed that it will be traversed from the head and bail out as early as possible, but e.g guaranteeing to traverse all elements would also be useful semantics, if you want to rely on the side effects. But for these functions I think you are happy with as fast as possible. You can not know beforehand if all elements will be traversed anyway, and making use of knowing that all before the true predicate in queue order are traversed feels far-fetched...

I'm ok with that, but let me point out that `queue:filter/2` is explicitly designed to operate in queue order. As the filter function is a predicate (with the additional capability of inserting/replacing), by @jhogberg's argument, this is equally not required and could be faster if the rear list was also just traversed as-is.

The traversal in queue order is not mentioned in the user documentation (only in the code), but it sets a precedent that strengthens the intuitive assumtion that traversal will be done in queue order in all functions.

I guess what I'm trying to say is, if the queue module is to go against the intuitive assumption, it should be mentioned in the documentation. That is, only mention that `queue:any` and `queue:all` do _not_ obey queue order, not that `filter` _does_.

Reply:
@RaimoNiskanen @jhogberg I will go with the approach that you suggested. While my approach _is_ faster than the suggested, it only matters in really extreme cases (some millions of items in a queue, likely a bad idea to begin with). Also, I cannot repeat my initial benchmarks that showed a significantly large difference, so I must have done something wrong back then (I didn't keep the setup, my bad :\)

Reply:
Obsolete, changed to `lists:all(Pred, F) andalso lists:all(Pred, R)` according to the discussion below, sacrificing traversal in queue order.